### ISSUE TITLE
Upgrading on old linux fails to detect correct version

### ISSUE THREAD

**Issue by mitom** [chef-cookbooks/chef_client_updater#73]
### Cookbook version
3.1.0

### Chef-client version
12.21.12

### Platform Details
Linux ip-10-244-6-170 3.2.0-23-virtual #36-Ubuntu SMP Tue Apr 10 22:29:03 UTC 2012 x86_64 x86_64 x86_64 GNU/Linux
AWS - Ubuntu 12.02

### Scenario:
Update chef to `~12`.

### Steps to Reproduce:
Run chef client with an update resource in a cookbook.

### Expected Result:
To detect that 12.21.12 is the latest for ubuntu 12.02

### Actual Result:
Tries to upgrade to 12.21.26 and assumes it succeeded.
```
    - upgrade chef-client 12.21.12 to 12.21.26
    ================================================================================
    Error executing action `update` on resource 'chef_client_updater[update chef-client]'
    ================================================================================

    SystemExit
    ----------
    exit

    Cookbook Trace:
    ---------------
    /var/chef/cache/cookbooks/chef_client_updater/providers/default.rb:176:in `exit'
    /var/chef/cache/cookbooks/chef_client_updater/providers/default.rb:176:in `run_post_install_action'
    /var/chef/cache/cookbooks/chef_client_updater/providers/default.rb:295:in `block (2 levels) in class_from_file'
    /var/chef/cache/cookbooks/chef_client_updater/providers/default.rb:294:in `block in class_from_file'

    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/chef_client_updater/recipes/default.rb

     20: chef_client_updater 'update chef-client' do
     21:   channel node['chef_client_updater']['channel']
     22:   version node['chef_client_updater']['version']
     23:   prevent_downgrade node['chef_client_updater']['prevent_downgrade']
     24:   post_install_action node['chef_client_updater']['post_install_action']
     25:   download_url_override node['chef_client_updater']['download_url_override'] if node['chef_client_updater']['download_url_override']
     26:   checksum node['chef_client_updater']['checksum'] if node['chef_client_updater']['checksum']
     27: end

    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/chef_client_updater/recipes/default.rb:20:in `from_file'

    chef_client_updater("update chef-client") do
      action [:update]
      updated true
      updated_by_last_action true
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      declared_type :chef_client_updater
      cookbook_name "chef_client_updater"
      recipe_name "default"
      channel "stable"
      version "12"
      prevent_downgrade false
      post_install_action "kill"
      exec_command "chef-client"
      exec_args []
    end

    System Info:
    ------------
    chef_version=12.21.12
    platform=ubuntu
    platform_version=12.04
    ruby=ruby 2.3.5p376 (2017-09-14 revision 59905) [x86_64-linux]
    program_name=chef-client worker: ppid=23536;start=16:45:55;
    executable=/opt/chef/bin/chef-client


Running handlers:
Running handlers complete
Chef Client failed. 1 resources updated in 23 seconds
```

Trying to upgrade manually it seems the latest for 12.02 is 12.21.12:
```
curl -L https://chef.io/chef/install.sh | sudo bash -s -- -v 12
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 20493  100 20493    0     0   135k      0 --:--:-- --:--:-- --:--:--  909k
ubuntu 12.04 x86_64
Getting information for chef stable 12 for ubuntu...
downloading http://www.chef.io/stable/chef/metadata?v=12&p=ubuntu&pv=12.04&m=x86_64
  to file /tmp/install.sh.25283/metadata.txt
trying wget...
sha1	85e4ba824c71e8c82a512edc97eaeb0236667f4f
sha256	d997a79c4bdaf02397c464a18b1738bb33cbcc939c0898536de22ceee63f200c
url	https://packages.chef.io/files/stable/chef/12.21.12/ubuntu/12.04/chef_12.21.12-1_amd64.deb
version	12.21.12
downloaded metadata file looks valid...
downloading https://packages.chef.io/files/stable/chef/12.21.12/ubuntu/12.04/chef_12.21.12-1_amd64.deb
  to file /tmp/install.sh.25283/chef_12.21.12-1_amd64.deb
trying wget...
```

While ubuntu 12.02 is past EOL and therefore shouldn't expect support, it would be nice if this didn't break chef on it by exiting without any errors.

**Comment by mitom** [comment]
I told it to upgrade to `12`, I'd assume it functions the same way - in terms of the logic selecting which version to install - as the manual command (the script, really): 
```
curl -L https://chef.io/chef/install.sh | sudo bash -s -- -v 12
```

The cookbook failed to detect the correct latest version for the constraint and tried to install an unsupported one which correctly failed, but it selected the wrong version itself. The script correctly detected the latest version for the same constraint on the same system, and then succeeded in installing it.

As I said I did not expect it to succeed, nor that it is supported given that it's EOL, but as you said the error message does not help at all. I did missphrase myself though. It would read better as:
>While ubuntu 12.02 is past EOL and therefore shouldn't expect support, it would be nice if this didn't break chef on it by exiting without ~~any errors~~ _an explanation for the cause of the failure_.
